cmd/key-hash-backfill/main.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"crypto/sha256"
+	"crypto/x509"
+	"database/sql"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/db"
+	"github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+func parseCert(certBytes []byte) ([]byte, error) {
+	cert, err := x509.ParseCertificate(certBytes)
+	if err != nil {
+		return nil, err
+	}
+	h := sha256.Sum256(cert.RawSubjectPublicKeyInfo)
+	return h[:], nil
+}
+
+func insertCert(dbMap *db.WrappedMap, keyHash []byte, expires time.Time, serial string) error {
+	_, err := dbMap.Exec(
+		`INSERT INTO keyHashToSerial
+		(keyHash, certExpires, certSerial)
+		VALUES (?, ?, ?)`,
+		keyHash,
+		expires,
+		serial,
+	)
+	return err
+}
+
+type workUnit struct {
+	Serial  string
+	DER     []byte
+	Expires time.Time
+}
+
+var batchSize = 1000
+
+func getWork(logger log.Logger, dbMap *db.WrappedMap) ([]workUnit, error) {
+	var data []workUnit
+	// Get rows that are in certificates but not in keyHashToSerial,
+	// this should trend towards 0
+	_, err := dbMap.Select(
+		&data,
+		`SELECT a.serial, a.der, a.expires
+		FROM certificates AS a
+		LEFT JOIN keyHashToSerial AS b
```

Comment:
I'm skeptical of doing a JOIN for this. We've found that JOINs have poor performance in the past.

Instead I'd recommend doing an iteration by `id` on the certificates table, starting from a user-selected id, and insert everything, using the `IGNORE` keyword to ignore duplicate inserts.